src/array.c

```diff
@@ -399,14 +410,37 @@ const void * const*array_get_buffer(Array *ar)
  */
 enum cc_stat array_index_of(Array *ar, void *element, size_t *index)
 {
-    size_t i;
-    for (i = 0; i < ar->size; i++) {
-        if (ar->buffer[i] == element) {
-            *index = i;
-            return CC_OK;
+    if(ar->is_sorted){
+        size_t left=0, right = ar->size-1, middle;
+	int cmp;
+	bool done=false;
+	while(!done){
+            middle = (right-left)/2 + left;
+	    cmp = ar->comp(element,ar->buffer[middle]);
+	    if(cmp < 0){
+                right = middle;
+	    }else if(cmp > 0){
+                left = middle;
+	    }else if(cmp == 0){
+                done = true;
+	    }
+	    if(left > right) return CC_ERR_OUT_OF_RANGE;
+	}
+	size_t i;
+	for(i = middle; ar->comp(element,ar->buffer[i]) == 0 && i>=0; i--);
```

Comment:
Actually the test  `i >= 0` is redundant as it is always true since `i` is an unsigned integer of `size_t`. Even if you go under 0 it would just overflow and still test positive.